### PR TITLE
HPCC-16179 Redundant LDAP server not used

### PR DESCRIPTION
**Pull request by RussWhitehead**
The LDAP connection pool incorrectly creates connections to all LDAP servers
listed in the configuration. This is bad because a resource could be created
on one LDAP, followed by attempts to reference it on another LDAP.
The correct behavior is to use the same one, and
if it fails move on to the next one. This PR replaces the old CLoadBalancer
class with a CHostManager, which implements the correct behavior

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

### CODE REVIEW

**Comment by RussWhitehead** [comment]
@rpastrana Please review

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 1]
```diff
@@ -317,24 +327,30 @@ class CLdapConfig : implements ILdapConfig, public CInterface
...
-            if(retries < LDAPSEC_MAX_RETRIES)
+            getLdapHost(hostbuf);
+            for(int retries = 0; retries <= LDAPSEC_MAX_RETRIES; retries++)
             {
-                DBGLOG("Server %s temporarily unreachable.", hostbuf.str());
-                // Retrying next ldap sever, might be the same server
-                hostbuf.clear();
-                getLdapHost(hostbuf);
-                DBGLOG("Retrying with %s...", hostbuf.str());
+                rc = LdapUtils::getServerInfo(hostbuf.str(), m_ldapport, dcbuf, m_serverType, cfg->queryProp(".//@ldapDomain"));
```
Comment (rpastrana) [inline system/security/LdapSecurity/ldapconnection.cpp]: is it worthwhile extracting cfg->queryProp(".//@ldapDomain") once and storing in local var?

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 2]
```diff
@@ -492,28 +508,22 @@ class CLdapConfig : implements ILdapConfig, public CInterface
...
-        catch(...)
-        {
-            DBGLOG("getLdapHost unknown exception");
-        }
+        return s_hostManager.queryNumHosts();
+    }
 
+    virtual StringBuffer& getLdapHost(StringBuffer& hostbuf)
+    {
+        hostbuf.clear().append(s_hostManager.queryCurrentHost());
```
Comment (rpastrana) [inline system/security/LdapSecurity/ldapconnection.cpp]: hostbuf.set(s_hostManager.queryCurrentHost()); is a little cleaner

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 3]
```diff
@@ -738,45 +748,53 @@ class CLdapConnection : implements ILdapConnection, public CInterface
...
-            hostbuf.clear();
             m_ldapconfig->getLdapHost(hostbuf);
-        }
 
-        if(rc == LDAP_SERVER_DOWN)
-        {
-            StringBuffer dc;
-            LdapUtils::getDcName(m_ldapconfig->getDomain(), dc);
-            if(dc.length() > 0)
+            const char* proto;
```
Comment (rpastrana) [inline system/security/LdapSecurity/ldapconnection.cpp]: we might not need to resolve the prototype multiple times, move out of loop

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 4]
```diff
@@ -738,45 +748,53 @@ class CLdapConnection : implements ILdapConnection, public CInterface
...
+                }
+            }
+
+            if (rc != LDAP_SUCCESS)
             {
-                WARNLOG("Using automatically obtained LDAP Server %s", dc.str());
-                rc = connect(dc.str(), proto);
+                m_ldapconfig->blacklistHost(hostbuf);
             }
+
```
Comment (rpastrana) [inline system/security/LdapSecurity/ldapconnection.cpp]: this loop is only iterating once due to this break;

**Review by rpastrana — APPROVED** [review]
@RussWhitehead left a few comments

**Comment by RussWhitehead** [comment]
@rpastrana Please re-review changes based on your recommendations

**Comment by rpastrana** [comment]
@RussWhitehead looks good

**Comment by RussWhitehead** [comment]
@richardkchapman Please merge into 6.2.0 and close

**Comment by RussWhitehead** [comment]
@richardkchapman Please merge

File: system/security/LdapSecurity/ldapconnection.hpp [hunk 5]
```diff
@@ -85,7 +85,11 @@ ldap_compare_ext_s LDAP_P((
     typedef struct timeval TIMEVAL;
 #endif
 
-#define LDAPTIMEOUT 60 //20 second connection/search timeout
+#ifdef _DEBUG
+	#define LDAPTIMEOUT 5
+#else
+	#define LDAPTIMEOUT 20
```
Comment (richardkchapman) [inline system/security/LdapSecurity/ldapconnection.hpp]: Why was this changed from 60 to 20 - was it intentional?

**Comment by richardkchapman** [comment]
Merged manually to squash the commits
